zilliqa/src/consensus.rs

```diff
@@ -1777,4 +1841,53 @@ impl Consensus {
 
         Ok(())
     }
+
+    /// Deal with the fork to this block. The block is assumed to be valid to switch to.
+    /// This will make it so the block is ready to become the new head
+    fn deal_with_fork(&mut self, block: &Block) -> Result<()> {
+        // Start at the head block and recursively unwind the changes that have been made
+        // until the head block is the parent of this block.
+        while self.head_block().header.number >= block.header.number {
+            let head_block = self.head_block();
+            let parent_block = self
+                .get_block(&head_block.parent_hash())?
+                .ok_or_else(|| anyhow!("missing block parent when reverting blocks!"))?;
+
+            if head_block.header.view == 0 {
```

Comment:
I don't think genesis should ever come into the equation here - proposals should always be height >= 1 so the parent is genesis (for the case of a fork at genesis), but if logic is correct it will never need to request genesis parent.

The check is just for database corruption/race conditions. If you start a chain unwind and you have no HEAD parent.

I guess maybe it could happen on a light node too now I think of it.

Reply:
You're probably right but then I don't think the `block.view == 0` check is necessary at all.
I think we are currently assuming that the database isn't going to be corrupted, because if it is then anything goes and every part of consensus is liable to failing. I'd remove this if-statement to simplify the code a bit and shorten the function.

Reply:
Yeah, makes sense 👍